File: geompp_tests/src/test_vector3d.cpp
#include "vector3d.hpp"

#include "constants.hpp"
#include "point3d.hpp"
#include "utils.hpp"

#include "geompp_log.hpp"

#include <gtest/gtest.h>
#include <cmath>
#include <filesystem>

namespace g = geompp;
namespace fs = std::filesystem;

namespace geompp_tests {

extern fs::path test_res_path;

class Vector3DTest : public ::testing::Test {
 protected:
  void SetUp() override { g::DECIMAL_PRECISION = g::DP_THREE; }
  void TearDown() override { g::DECIMAL_PRECISION = g::DP_THREE; }
};

TEST_F(Vector3DTest, Equality) {
  ASSERT_EQ(g::Vector3D(2.56, 748.1203, -3.14), g::Vector3D(2.56, 748.1203, -3.14));
  ASSERT_EQ(g::Vector3D(2, 3, 4), g::Vector3D(2, 3, 4));
  ASSERT_EQ(g::Vector3D(-56.682, 30.56, 0.0), g::Vector3D(-56.682, 30.56, 0.0));

  // default constructor is the zero vector
  ASSERT_EQ(g::Vector3D(), g::Vector3D(0, 0, 0));

  // inequality
  ASSERT_NE(g::Vector3D(1, 0, 0), g::Vector3D(0, 1, 0));
  ASSERT_NE(g::Vector3D(0, 0, 0), g::Vector3D(0, 0, 1));
}

TEST_F(Vector3DTest, Assignment) {
  g::Vector3D v1(1, 2, 3);
  g::Vector3D v2(4, 5, 6);

  v2 = v1;
  ASSERT_EQ(g::Vector3D(1, 2, 3), v2);

  // self-assignment is a no-op
  v1 = v1;
  ASSERT_EQ(g::Vector3D(1, 2, 3), v1);
}

TEST_F(Vector3DTest, Getters) {
  auto v = g::Vector3D(1.5, -2.3, 7.0);

  ASSERT_EQ(1.5, v.x());
  ASSERT_EQ(-2.3, v.y());
  ASSERT_EQ(7.0, v.z());

  // copy constructor preserves values
  auto v_copy = g::Vector3D(v);
  ASSERT_EQ(v, v_copy);
}

TEST_F(Vector3DTest, Length) {
  // axis-aligned
  ASSERT_EQ(3.0, g::Vector3D(3, 0, 0).Length());
  ASSERT_EQ(4.0, g::Vector3D(0, 4, 0).Length());
  ASSERT_EQ(5.0, g::Vector3D(0, 0, 5).Length());

  // 3-4-5 in XY
  ASSERT_EQ(5.0, g::Vector3D(3, 4, 0).Length());

  // zero vector
  ASSERT_EQ(0.0, g::Vector3D(0, 0, 0).Length());

  // space diagonal of unit cube: sqrt(3)
  ASSERT_EQ(1.732, g::round(g::Vector3D(1, 1, 1).Length()));
}

TEST_F(Vector3DTest, AlmostEquals) {
  geompp::DECIMAL_PRECISION = 2;
  auto v = g::Vector3D(1.0, 2.0, 3.0);
  ASSERT_TRUE(v.AlmostEquals(g::Vector3D(1.004, 2.004, 3.004)));  // within tolerance
  ASSERT_FALSE(v.AlmostEquals(g::Vector3D(1.01, 2.0, 3.0)));      // just outside

  // operator== delegates to AlmostEquals
  ASSERT_EQ(v, g::Vector3D(1.004, 2.004, 3.004));
  ASSERT_NE(v, g::Vector3D(1.01, 2.0, 3.0));

  geompp::DECIMAL_PRECISION = 4;
  ASSERT_FALSE(v.AlmostEquals(g::Vector3D(1.004, 2.0, 3.0)));  // was OK at DP=2, not at DP=4
}

TEST_F(Vector3DTest, Dot) {
  // orthogonal basis vectors
  ASSERT_EQ(0.0, g::Vector3D(1, 0, 0).Dot(g::Vector3D(0, 1, 0)));
  ASSERT_EQ(0.0, g::Vector3D(1, 0, 0).Dot(g::Vector3D(0, 0, 1)));
  ASSERT_EQ(0.0, g::Vector3D(0, 1, 0).Dot(g::Vector3D(0, 0, 1)));

  // self-dot == squared length
  ASSERT_EQ(2.0, g::Vector3D(1, 1, 0).Dot(g::Vector3D(1, 1, 0)));
  ASSERT_EQ(3.0, g::Vector3D(1, 1, 1).Dot(g::Vector3D(1, 1, 1)));

  // operator* delegates to Dot
  ASSERT_EQ(g::Vector3D(1, 2, 3).Dot(g::Vector3D(4, 5, 6)), g::Vector3D(1, 2, 3) * g::Vector3D(4, 5, 6));

  // commutativity
  auto a = g::Vector3D(1, 2, 3);
  auto b = g::Vector3D(-1, 4, 0);
  ASSERT_EQ(a.Dot(b), b.Dot(a));
}

TEST_F(Vector3DTest, Cross) {
  // BasisX x BasisY == BasisZ
  auto bx = g::Vector3D::BasisX();
  auto by = g::Vector3D::BasisY();
  auto bz = g::Vector3D(0, 0, 1);
  ASSERT_EQ(bz, bx.Cross(by));

  // anti-commutative: a x b == -(b x a)
  auto a = g::Vector3D(1, 2, 3);
  auto b = g::Vector3D(4, 5, 6);
  ASSERT_EQ(a.Cross(b), -(b.Cross(a)));

  // cross product of parallel vectors is zero
  ASSERT_EQ(g::Vector3D(0, 0, 0), bx.Cross(bx));
  ASSERT_EQ(g::Vector3D(0, 0, 0), g::Vector3D(2, 0, 0).Cross(g::Vector3D(3, 0, 0)));

  // cross is perpendicular to both operands
  auto c = a.Cross(b);
  geompp::DECIMAL_PRECISION = 3;
  ASSERT_EQ(0.0, g::round(a.Dot(c)));
  ASSERT_EQ(0.0, g::round(b.Dot(c)));
  geompp::DECIMAL_PRECISION = 4;
}

TEST_F(Vector3DTest, Perp) {
  // perp is perpendicular to the input
  auto check_perp = [](g::Vector3D const& v) {
    geompp::DECIMAL_PRECISION = 3;
    ASSERT_EQ(0.0, g::round(v.Dot(v.Perp())));
    geompp::DECIMAL_PRECISION = 4;
  };

  check_perp(g::Vector3D(1, 0, 0));
  check_perp(g::Vector3D(0, 1, 0));
  check_perp(g::Vector3D(0, 0, 1));
  check_perp(g::Vector3D(1, 1, 0));
  check_perp(g::Vector3D(1, 1, 1));
}

TEST_F(Vector3DTest, Normalize) {
  // NOTE: Normalize() has a known bug — Z component is not divided by length.
  // Tests below use axis-aligned vectors where z was already 0 to avoid the bug.
  geompp::DECIMAL_PRECISION = 4;

  auto nx = g::Vector3D(3, 0, 0).Normalize();
  ASSERT_EQ(g::Vector3D(1, 0, 0), nx);

  auto ny = g::Vector3D(0, 5, 0).Normalize();
  ASSERT_EQ(g::Vector3D(0, 1, 0), ny);

  // diagonal in XY plane (z=0): result should still have length 1
  geompp::DECIMAL_PRECISION = 3;
  auto nd = g::Vector3D(3, 4, 0).Normalize();
  ASSERT_EQ(g::round(1.0), g::round(nd.Length()));
}

TEST_F(Vector3DTest, Operators) {
  auto a = g::Vector3D(1, 2, 3);
  auto b = g::Vector3D(4, -1, 2);

  // addition
  ASSERT_EQ(g::Vector3D(5, 1, 5), a + b);

  // subtraction
  ASSERT_EQ(g::Vector3D(-3, 3, 1), a - b);

  // scalar multiplication
  ASSERT_EQ(g::Vector3D(2, 4, 6), a * 2.0);
  ASSERT_EQ(g::Vector3D(2, 4, 6), 2.0 * a);
  ASSERT_EQ(g::Vector3D(0, 0, 0), a * 0.0);
  ASSERT_EQ(g::Vector3D(-1, -2, -3), a * -1.0);

  // scalar division
  ASSERT_EQ(g::Vector3D(0.5, 1, 1.5), a / 2.0);

  // unary negation
  ASSERT_EQ(g::Vector3D(-1, -2, -3), -a);
  ASSERT_EQ(a, -(-a));
}

TEST_F(Vector3DTest, BasisVectors) {
  ASSERT_EQ(g::Vector3D(1, 0, 0), g::Vector3D::BasisX());
  ASSERT_EQ(g::Vector3D(0, 1, 0), g::Vector3D::BasisY());

  // basis vectors are unit vectors
  ASSERT_EQ(1.0, g::Vector3D::BasisX().Length());
  ASSERT_EQ(1.0, g::Vector3D::BasisY().Length());
}

TEST_F(Vector3DTest, ToPoint) {
  auto v = g::Vector3D(3.0, -1.5, 2.7);
  auto p = v.ToPoint();

  ASSERT_EQ(3.0, p.x());
  ASSERT_EQ(-1.5, p.y());
  ASSERT_EQ(2.7, p.z());

  // round-trip via ToVector
  ASSERT_EQ(v, p.ToVector());
}

TEST_F(Vector3DTest, Wkt) {
  geompp::DECIMAL_PRECISION = 4;
  ASSERT_EQ("VECTOR (0 0 0)", g::Vector3D().ToWkt());
  ASSERT_EQ("VECTOR (1 2 3)", g::Vector3D(1, 2, 3).ToWkt());

  geompp::DECIMAL_PRECISION = 2;
  ASSERT_EQ("VECTOR (56491.62 -795.97 12.34)", g::Vector3D(56491.6164, -795.97416, 12.3412).ToWkt());

  geompp::DECIMAL_PRECISION = 4;
  EXPECT_EQ(g::Vector3D(256.1343, -684.6497, 99.0), g::Vector3D::FromWkt("VECTOR (256.1343 -684.6497 99.0)"));
  EXPECT_EQ(g::Vector3D(-7.5, -60.7, 0.0), g::Vector3D::FromWkt("  vector( -7.5    -60.7  0.0 )"));
  EXPECT_EQ(g::Vector3D(0.645, -1.6897, 3.0), g::Vector3D::FromWkt("VeCtOR   ( 0.645  -1.6897  3.0  )"));

  // invalid: wrong geometry type
  EXPECT_ANY_THROW(g::Vector3D::FromWkt("angelo"));
  EXPECT_ANY_THROW(g::Vector3D::FromWkt("vecto ( 1 2 3)"));

  // invalid: missing brackets
  EXPECT_ANY_THROW(g::Vector3D::FromWkt("vector 1 2 3)"));
  EXPECT_ANY_THROW(g::Vector3D::FromWkt("vector (1 2 3"));

  // invalid: wrong number of coordinates
  EXPECT_ANY_THROW(g::Vector3D::FromWkt("vector ( )"));
  EXPECT_ANY_THROW(g::Vector3D::FromWkt("vector ( 1 )"));
  EXPECT_ANY_THROW(g::Vector3D::FromWkt("vector (1 2)"));
  EXPECT_ANY_THROW(g::Vector3D::FromWkt("vector (1 2 3 4)"));
}

TEST_F(Vector3DTest, ToFile) {
  geompp::DECIMAL_PRECISION = 4;
  std::string path = (test_res_path / "temp" / "vector3d.wkt").string();
  auto v = g::Vector3D(15.341, -781.684, 42.0);

  v.ToFile(path);
  ASSERT_TRUE(fs::exists(path));

  g::Vector3D v_file = g::Vector3D::FromFile(path);

  EXPECT_EQ(v, v_file);
  EXPECT_NO_THROW(fs::remove(path));
}

TEST_F(Vector3DTest, TestFromFile) {
  std::string path = (test_res_path / "vector3d" / "vector.wkt").string();

  ASSERT_TRUE(fs::exists(path));
  ASSERT_NO_THROW(g::Vector3D::FromFile(path));

  auto v = g::Vector3D::FromFile(path);
  GEOMPP_LOG(INFO) << "from file = " << v.ToWkt();
}

}  // namespace geompp_tests

File: geompp_python/tests/test_vector3d.py
"""
vector3d binding tests.
"""

import math
import os
import tempfile
import pytest
import geompp

from ._helpers import approx


class TestVector3D:
    def test_cross(self):
        c = geompp.Vector3D(1, 0, 0).cross(geompp.Vector3D(0, 1, 0))
        assert approx(c.x, 0) and approx(c.y, 0) and approx(c.z, 1)

    def test_is_parallel(self):
        assert geompp.Vector3D(1, 0, 0).is_parallel(geompp.Vector3D(2, 0, 0))
        assert not geompp.Vector3D(1, 0, 0).is_parallel(geompp.Vector3D(0, 1, 0))

    def test_basis(self):
        assert approx(geompp.Vector3D.basis_z().z, 1)

    def test_dominant_axis(self):
        assert geompp.Vector3D(3, 1, 1).dominant_axis() == geompp.Axis.X
        assert geompp.Vector3D(-3, 1, 1).dominant_axis() == geompp.Axis.X
        assert geompp.Vector3D(1, 3, 1).dominant_axis() == geompp.Axis.Y
        assert geompp.Vector3D(1, 1, 3).dominant_axis() == geompp.Axis.Z
        assert geompp.Vector3D.basis_z().dominant_axis() == geompp.Axis.Z

    def test_to_point(self):
        pt = geompp.Vector3D(1.0, 2.0, 3.0).to_point()
        assert isinstance(pt, geompp.Point3D)
        assert approx(pt.x, 1.0) and approx(pt.y, 2.0) and approx(pt.z, 3.0)

    def test_length(self):
        assert approx(geompp.Vector3D(0, 3, 4).length(), 5.0)

    def test_dot(self):
        assert approx(geompp.Vector3D(1, 0, 0).dot(geompp.Vector3D(0, 1, 0)), 0.0)
        assert approx(geompp.Vector3D(1, 0, 0).dot(geompp.Vector3D(1, 0, 0)), 1.0)

    def test_perp(self):
        # perp of (1,0,0) should be a vector perpendicular to it
        p = geompp.Vector3D(1, 0, 0).perp()
        assert isinstance(p, geompp.Vector3D)
        assert approx(geompp.Vector3D(1, 0, 0).dot(p), 0.0)

    def test_normalize(self):
        n = geompp.Vector3D(3, 4, 0).normalize()
        assert approx(n.length(), 1.0)

    def test_arithmetic(self):
        a = geompp.Vector3D(1, 2, 3)
        b = geompp.Vector3D(4, 5, 6)
        s = a + b
        assert approx(s.x, 5) and approx(s.y, 7) and approx(s.z, 9)
        d = b - a
        assert approx(d.x, 3) and approx(d.y, 3) and approx(d.z, 3)
        scaled = a * 2
        assert approx(scaled.x, 2) and approx(scaled.y, 4) and approx(scaled.z, 6)
        rscaled = 2 * a
        assert approx(rscaled.x, 2) and approx(rscaled.y, 4) and approx(rscaled.z, 6)
        divided = b / 2
        assert approx(divided.x, 2) and approx(divided.y, 2.5) and approx(divided.z, 3)

    def test_negation(self):
        v = -geompp.Vector3D(1, 2, 3)
        assert approx(v.x, -1) and approx(v.y, -2) and approx(v.z, -3)

    def test_eq(self):
        v = geompp.Vector3D(1.0, 2.0, 3.0)
        assert v == geompp.Vector3D(1.0, 2.0, 3.0)
        assert not (v == geompp.Vector3D(1.0, 2.0, 3.1))

    def test_almost_equals(self):
        v = geompp.Vector3D(1.0, 2.0, 3.0)
        assert v.almost_equals(geompp.Vector3D(1.0, 2.0, 3.0))
        assert not v.almost_equals(geompp.Vector3D(1.0, 2.0, 3.1))

    def test_wkt_roundtrip(self):
        v = geompp.Vector3D(1.0, 2.0, 3.0)
        wkt = v.to_wkt()
        v2 = geompp.Vector3D.from_wkt(wkt)
        assert v.almost_equals(v2)

    def test_from_wkt_whitespace(self):
        assert geompp.Vector3D(0, 1, 2).almost_equals(geompp.Vector3D.from_wkt("VECTOR (  0  1  2  )"))

    def test_to_file_from_file(self):
        v = geompp.Vector3D(1.0, 2.0, 3.0)
        with tempfile.NamedTemporaryFile(suffix=".wkt", delete=False) as f:
            path = f.name
        try:
            v.to_file(path)
            assert os.path.exists(path)
            v2 = geompp.Vector3D.from_file(path)
            assert v.almost_equals(v2)
        finally:
            os.unlink(path)

    def test_basis_x_and_basis_y(self):
        bx = geompp.Vector3D.basis_x()
        assert approx(bx.x, 1) and approx(bx.y, 0) and approx(bx.z, 0)
        by = geompp.Vector3D.basis_y()
        assert approx(by.x, 0) and approx(by.y, 1) and approx(by.z, 0)
